### PR TITLE
ePay payment provider, MD5, callback

### PR DESCRIPTION
**Pull request by Ownoo**
ePay payment provider reworked to do MD5 checksum if "secret" present, as well as splitting callback and success urls to provide safe server-to-server callback even if cardholder aborts after payment window. With error logging to default log.

### CODE REVIEW

File: uWebshop.Payment.ePay/ePayPaymentBase.cs [hunk 1]
```diff
@@ -1,10 +1,29 @@
...
+        public static string MD5(string phrase)
+        {
+            var hasher = System.Security.Cryptography.MD5.Create();
+            var hashedDataBytes = hasher.ComputeHash(Encoding.UTF8.GetBytes(phrase));
+            return ByteArrayToString(hashedDataBytes);
+        }
+
+        private static string ByteArrayToString(byte[] inputArray)
+        {
+            var sb = new StringBuilder();
```
Comment (kappy) [inline uWebshop.Payment.ePay/ePayPaymentBase.cs]: you know beforehand the size of the inputArray you want to convert to string, so you know the size of the returning string (2*inputArray.Lenght). You initialize StringBuilder capacity with that value, you save one internal reallocation of it.

File: uWebshop.Payment.ePay/ePayPaymentRequestHandler.cs [hunk 2]
```diff
@@ -1,63 +1,66 @@
...
+            request.Parameters.Add("currency", orderInfo.StoreInfo.Store.CurrencyCultureSymbol);
+            request.Parameters.Add("windowstate", "3");
+
+            //build MD5 if secret present
+            if (secret != string.Empty)
+            {
+                var values = string.Empty;
+                foreach (var parameter in request.Parameters)
+                {
+                    values += parameter.Value;
```
Comment (kappy) [inline uWebshop.Payment.ePay/ePayPaymentRequestHandler.cs]: This is where the StringBuilder should really be applied. a String concatenation inside a foreach cycle.

File: uWebshop.Payment.ePay/ePayPaymentResponseHandler.cs [hunk 3]
```diff
@@ -6,45 +6,64 @@
...
+                var amount = HttpContext.Current.Request.QueryString["amount"] ?? "0";
+                var validated = true;
+                if (secret != string.Empty)
+                {
+                    var values = string.Empty;
+                    foreach (var key in HttpContext.Current.Request.QueryString.AllKeys)
+                    {
+                        if (key != "hash")
+                        {
+                            values += HttpContext.Current.Request.QueryString[key];
```
Comment (kappy) [inline uWebshop.Payment.ePay/ePayPaymentResponseHandler.cs]: Here again

File: uWebshop.Payment.ePay/ePayPaymentBase.cs (hunk 1 above)
Comment (Ownoo) [inline uWebshop.Payment.ePay/ePayPaymentBase.cs]: I agree. I will fix the StringBuilder issues.

**Pull request closed without merging**